clippy_lints/src/manual_unwrap_or.rs

```diff
@@ -78,24 +96,37 @@ fn lint_option_unwrap_or_case<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tc
     if_chain! {
         if let ExprKind::Match(scrutinee, match_arms, _) = expr.kind;
         let ty = cx.typeck_results().expr_ty(scrutinee);
-        if utils::is_type_diagnostic_item(cx, ty, sym!(option_type));
-        if let Some(none_arm) = applicable_none_arm(match_arms);
+        if let Some(case) = if utils::is_type_diagnostic_item(cx, ty, sym!(option_type)) {
+            Some(Case::Option)
+        } else if utils::is_type_diagnostic_item(cx, ty, sym!(result_type)) {
+            Some(Case::Result)
+        } else {
+            None
+        };
+        if let Some(or_arm) = applicable_or_arm(match_arms);
         if let Some(scrutinee_snippet) = utils::snippet_opt(cx, scrutinee.span);
-        if let Some(none_body_snippet) = utils::snippet_opt(cx, none_arm.body.span);
+        if let Some(or_body_snippet) = utils::snippet_opt(cx, or_arm.body.span);
         if let Some(indent) = utils::indent_of(cx, expr.span);
-        if constant_simple(cx, cx.typeck_results(), none_arm.body).is_some();
+        if constant_simple(cx, cx.typeck_results(), or_arm.body).is_some();
         then {
-            let reindented_none_body =
-                utils::reindent_multiline(none_body_snippet.into(), true, Some(indent));
+            let reindented_or_body =
+                utils::reindent_multiline(or_body_snippet.into(), true, Some(indent));
+            let wrap_in_parens = !matches!(scrutinee, Expr {
+                kind: ExprKind::Call(..) | ExprKind::Path(_), ..
```

Comment:
With this approach we will miss e.g. method calls
```rust
struct S {}
impl S {
    fn method(self) -> Option<i32> {
        Some(42)
    }
}

fn main() {
    let s = S {};
    match s.method() {
        Some(i) => i,
        None => 42,
    };
}
```
The method call would be parenthesized.

I think this can be solved using the `Sugg` struct in `utils::sugg`, changing the suggestion like this:
```rust
Sugg::hir(cx, scrutinee, "..").maybe_par()
```
Sorry for not seeing this before

Reply:
Nice I didn't know this util :+1: